exonum/src/storage/proof_map_index/key.rs

```diff
@@ -462,6 +468,53 @@ impl PartialOrd for ProofPath {
     }
 }
 
+impl ProofPath {
+    /// Creates a compressed binary representation of the given proof path.
+    ///
+    /// # Binary format
+    ///
+    /// - **bits_len** - total length in bits compressed by the leb128 algorithm.
+    /// - **bytes** - non-null bytes of the given ProofPath, i.e. the first `(bits_len + 7) / 8` bytes.
+    pub fn compress(&self) -> SmallVec<[u8; 64]> {
+        let bits_len = self.end() as u64;
+        let bytes_len = ((bits_len + 7) / 8) as usize;
+        let key = &self.raw_key()[0..bytes_len];
+
+        let mut buf = smallvec![0u8; 64];
+        let bytes_written = {
+            let mut writer = Cursor::new(buf.as_mut());
+            let mut bytes_written = leb128::write::unsigned(&mut writer, bits_len).unwrap();
+            bytes_written + writer.write(&key).unwrap()
+        };
+        buf.resize(bytes_written, 0);
+        // Cuts of the bits that lie to the right of the end.
+        let last_bits_len = bits_len % 8;
+        if bytes_len > 0 && last_bits_len != 0 {
+            *buf.as_mut().last_mut().unwrap() &= !(255_u8 << (self.end() % 8));
```

Comment:
I definetely don't like that kind of one-liners. Can you explain what's going on here? (better in code comment)

Reply:
Initially I thought this masks off the trailing bits which are not zero for some reason.

Supposed the bit stream contains 13 bits (`bits_len == 13`) which is represented with the following two bytes:

```
0110001011101: stream

01000110 XXX10111: buf
```

then we would like to ensure that trailing bits (marked with X here) are actually zero. However, what we _seem_ to do here is this:

```
01000110 XXX10111: buf
         11111000: 255u8 << (bits_len % 8)
         00000111: !(255_u8 << (bits_len % 8))
01000110 00000111: buf.last_mut() &= !(255_u8 << (bits_len % 8))
```

First, I do not understand why we need to zero them out explicitly. Isn't `self.raw_key()` already property constructed?

Second, I think the correct expression would be 

```suggestion
            *buf.as_mut().last_mut().unwrap() &= 255_u8 >> (self.end() % 8);
```

Reply:
> Isn't self.raw_key() already property constructed?

Depends on what you mean by "properly" if a key is constructed as a prefix of another key, the bits are copied from the parent key and not zeroed out after the end of the prefix (for efficiency reasons, seemingly, although I'm not sure we ever  assessed performance impact).